### ISSUE TITLE
Retry on different resource

### ISSUE THREAD

**Issue by tczhao** [argoproj/argo-workflows#5471]
We want achieve a state where a pod gets rerun with a different resource request when the pod failed/oomed (similar to #4493)

We can achieve this through the use of "enhanced dag dependency" but this can be quite cumbersome to write.

Would it be a good idea to add something like podSpecPatch in the retryStrategy? or we are happy to leave it as it is using dag and use things like jsonnet to build the workflow yaml?

E.g. instead of

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: retry-on-spec-
  namespace: default
spec:
  entrypoint: diamond
  templates:
    - name: diamond
      dag:
        tasks:
          - name: A1
            template: A-spec1
          - name: A2
            depends: "A1.Failed"
            template: A-spec2
          - name: B
            depends: "A1.Succeeded || A2.Succeeded"
            template: B
    - name: A-spec1
      container:
        image: alpine:3.7
        command:
          - sh
          - -c
          - exit 1
    - name: A-spec2
      container:
        image: alpine:3.7
        command:
          - sh
          - -c
          - exit 0
    - name: B
      container:
        image: alpine:3.7
        command:
          - sh
          - -c
          - exit 0
```

we do something like

```
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: retry-on-spec-
  namespace: default
spec:
  entrypoint: diamond
  templates:
    - name: diamond
      dag:
        tasks:
          - name: A1
            template: A-spec1
          - name: B
            depends: "A1.Succeeded"
            template: B
    - name: A
      podSpecPatch: '{"containers":[{"name":"main", "resources":{"limits":{"cpu": "1" }}}]}'
      retryStrategy:
        limit: "2"
        retrySpec:
       	  - podSpecPatch: '{"containers":[{"name":"main", "resources":{"limits":{"cpu": "3" }}}]}'
       	  - podSpecPatch: '{"containers":[{"name":"main", "resources":{"limits":{"cpu": "5" }}}]}'
      container:
        image: alpine:3.7
        command:
          - sh
          - -c
          - exit 1
    - name: B
      container:
        image: alpine:3.7
        command:
          - sh
          - -c
          - exit 0
```

---

**Message from the maintainers**:

Impacted by this bug? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by simster7** [comment]
I think the correct approach to this would be to use the `{{retries}}` variable in conjunction with a variable expression. Something like this:

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  generateName: retry-add-memory-
spec:
  entrypoint: retry-add-memory
  templates:
  - name: retry-add-memory
    retryStrategy:
      limit: "5"
      retryPolicy: "Always"
    podSpecPatch: |
      containers:
        - name: main
          resources:
            limits:
              memory: "{{=(sprig.int(retries) + 1) * 64}}Mi"
    container:
      image: python
      command: ["python", "-c"]
      args: ["[i for i in range(10 ** 7)]"]
      resources:
        limits:
          memory: 64Mi
```

I believe, however, that variable expressions will be available in `v3.1` (we're currently in v3.0 RC). They should be on `latest` though if you want to start playing with them

**Comment by tczhao** [comment]
Thanks, this helps a lot
